File: designsystem/src/androidMain/kotlin/me/bookk/designsystem/theme/color/Colors.kt
package me.bookk.designsystem.theme.color

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import dev.icerock.moko.resources.ColorResource
import dev.icerock.moko.resources.compose.colorResource
import me.bookk.designsystem.resources.DesignSystem


object AppColors {

    val White: Color
        @Composable
        get() = colorResource(resource = DesignSystem.colors.white)
    val Black: Color
        @Composable
        get() = colorResource(resource = DesignSystem.colors.black)


    val LightScheme: AppColorScheme
        @Composable
        get() = AppColorScheme(
            background = colorResource(resource = DesignSystem.colors.lightBackground),
            primaryText = colorResource(resource = DesignSystem.colors.lightPrimaryText),
            secondaryText = colorResource(resource = DesignSystem.colors.lightSecondaryText),
            hintText = colorResource(resource = DesignSystem.colors.lightHintText),
            elevated = colorResource(resource = DesignSystem.colors.lightElevated),
            divider = colorResource(resource = DesignSystem.colors.lightDivider),
            actionText = colorResource(resource = DesignSystem.colors.lightActionText),
            actionTextDisabled = colorResource(resource = DesignSystem.colors.lightActionTextDisabled),
            inactive = colorResource(resource = DesignSystem.colors.lightInactive),
            blurredBackground = colorResource(resource = DesignSystem.colors.lightBlurredBackground),
            header = colorResource(resource = DesignSystem.colors.lightHeader),
            success = colorResource(resource = DesignSystem.colors.lightSuccess),
            inactiveToggle = colorResource(resource = DesignSystem.colors.lightInactiveToggle),
            buttonPrimary = colorResource(resource = DesignSystem.colors.lightButton),
            buttonActive = colorResource(resource = DesignSystem.colors.lightButtonActive),
            buttonInactive = colorResource(resource = DesignSystem.colors.lightButtonInactive),
            error = colorResource(resource = DesignSystem.colors.error)
        )

    val DarkScheme: AppColorScheme
        @Composable
        get() = AppColorScheme(
            background = colorResource(resource = DesignSystem.colors.darkBackground),
            primaryText = colorResource(resource = DesignSystem.colors.darkPrimaryText),
            secondaryText = colorResource(resource = DesignSystem.colors.darkSecondaryText),
            hintText = colorResource(resource = DesignSystem.colors.darkHintText),
            elevated = colorResource(resource = DesignSystem.colors.darkElevated),
            divider = colorResource(resource = DesignSystem.colors.darkDivider),
            actionText = colorResource(resource = DesignSystem.colors.darkActionText),
            actionTextDisabled = colorResource(resource = DesignSystem.colors.darkActionTextDisabled),
            inactive = colorResource(resource = DesignSystem.colors.darkInactive),
            blurredBackground = colorResource(resource = DesignSystem.colors.darkBlurredBackground),
            header = colorResource(resource = DesignSystem.colors.darkHeader),
            success = colorResource(resource = DesignSystem.colors.darkSuccess),
            inactiveToggle = colorResource(resource = DesignSystem.colors.darkInactiveToggle),
            buttonPrimary = colorResource(resource = DesignSystem.colors.darkButton),
            buttonActive = colorResource(resource = DesignSystem.colors.darkButtonActive),
            buttonInactive = colorResource(resource = DesignSystem.colors.darkButtonInactive),
            error = colorResource(resource = DesignSystem.colors.error)
        )
}

@Composable
fun animateResource(resource: ColorResource): Color {
    return animateColor(colorResource(resource = resource))
}

@Composable
fun animateColor(color: Color): Color {
    return animateColorAsState(color, animationSpec = spring(stiffness = Spring.StiffnessLow)).value
}

File: iosApp/iosApp/DesignSystem/Colors.swift
//
//  Colors.swift
//  iosApp
//
//  Created by Antiufieiev Michael on 29.12.2024.
//  Copyright © 2024 BookkMe. All rights reserved.
//

import shared
import SwiftUI


class AppColors {
    
    static var background: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkBackground,
                light: DesignSystem.colors().lightBackground
            )
        }
    }
    static var primary: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkPrimaryText,
                light: DesignSystem.colors().lightPrimaryText
            )
        }
    }
    static var hintText: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkHintText,
                light: DesignSystem.colors().lightHintText
            )
        }
    }
    static var elevated: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkElevated,
                light: DesignSystem.colors().lightElevated
            )
        }
    }
    static var actionText: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkActionText,
                light: DesignSystem.colors().lightActionText
            )
        }
    }
    static var actionTextDisabled: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkActionTextDisabled,
                light: DesignSystem.colors().lightActionTextDisabled
            )
        }
    }
    static var inactive: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkInactive,
                light: DesignSystem.colors().lightInactive
            )
        }
    }
    static var header: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkHeader,
                light: DesignSystem.colors().lightHeader
            )
        }
    }
    static var success: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkSuccess,
                light: DesignSystem.colors().lightSuccess
            )
        }
    }
    static var inactiveToggle: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkInactiveToggle,
                light: DesignSystem.colors().lightInactiveToggle
            )
        }
    }
    static var buttonPrimary: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkButton,
                light: DesignSystem.colors().lightButton
            )
        }
    }
    static var buttonActive: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkButtonActive,
                light: DesignSystem.colors().lightButtonActive
            )
        }
    }
    static var buttonInactive: Color {
        get {
            createColor(
                dark: DesignSystem.colors().darkButtonInactive,
                light: DesignSystem.colors().lightButtonInactive
            )
        }
    }
    static var error: Color {
        get {
            createColor(
                dark: DesignSystem.colors().error,
                light: DesignSystem.colors().error
            )
        }
    }
    
    static func createColor(dark:shared.ColorResource, light:shared.ColorResource) -> Color {
        return Color(
            UIColor { traits in
                traits.userInterfaceStyle == .dark ? dark.getUIColor() : light.getUIColor()
            }
        )
    }
}
